src/util/requests/RequestBuilder.cpp

```diff
@@ -0,0 +1,205 @@
+            return Unexpected{std::move(streamData.error())};
+
+        if (!SSL_set_tlsext_host_name(streamData->stream.native_handle(), host_.c_str())) {
+            beast::error_code errorCode;
+            errorCode.assign(static_cast<int>(::ERR_get_error()), asio::error::get_ssl_category());
+            return Unexpected{RequestError{"SSL setup failed", errorCode}};
+        }
+        return doRequestImpl(std::move(streamData.value()), yield, method);
+    }
+
+    auto streamData = impl::TcpStreamData{yield};
+    return doRequestImpl(std::move(streamData), yield, method);
+}
+
+template <typename StreamDataType>
+Expected<std::string, RequestError>
+RequestBuilder::doRequestImpl(StreamDataType&& streamData, asio::yield_context yield, http::verb const method)
+{
+    auto executor = asio::get_associated_executor(yield);
+
+    // Look up the domain name
+    beast::error_code errorCode;
+    tcp::resolver resolver(executor);
+    auto const resolverResult = resolver.async_resolve(host_, port_, yield[errorCode]);
+    if (errorCode)
+        return Unexpected{RequestError{"Resolve error", errorCode}};
+
+    auto& stream = streamData.stream;
+
+    // Make the connection on the IP address we get from a lookup
+    beast::get_lowest_layer(stream).expires_after(timeout_);
+    beast::get_lowest_layer(stream).async_connect(resolverResult, yield[errorCode]);
+    if (errorCode)
+        return Unexpected{RequestError{"Connection error", errorCode}};
+
+    // Set up HTTP method
+    request_.method(method);
+
+    // Perform SSL handshake
+    if constexpr (StreamDataType::sslEnabled) {
+        beast::get_lowest_layer(stream).expires_after(timeout_);
+        stream.async_handshake(asio::ssl::stream_base::client, yield[errorCode]);
+        if (errorCode)
+            return Unexpected{RequestError{"Handshake error", errorCode}};
+    }
+
+    // Send the HTTP request to the remote host
+    beast::get_lowest_layer(stream).expires_after(timeout_);
+    http::async_write(stream, request_, yield[errorCode]);
+    if (errorCode)
+        return Unexpected{RequestError{"Write error", errorCode}};
+
+    // This buffer is used for reading and must be persisted
+    beast::flat_buffer buffer;
+    // Declare a container to hold the response
+    http::response<http::string_body> response;
+
+    // Receive the HTTP response
+    http::async_read(stream, buffer, response, yield[errorCode]);
+    if (errorCode)
+        return Unexpected{RequestError{"Read error", errorCode}};
+
+    // Gracefully close the socket
+    beast::get_lowest_layer(stream).socket().shutdown(tcp::socket::shutdown_both, errorCode);
+
+    // not_connected happens sometimes
+    // so don't bother reporting it.
+    //
+    if (errorCode && errorCode != beast::errc::not_connected)
+        return Unexpected{RequestError{"Shutdown socket error", errorCode}};
+
+    if (response.result() != http::status::ok)
+        return Unexpected{RequestError{"Response status not OK"}};
```

Comment:
So we report the response error only if we close the connection successfully?

Reply:
I can't imagine what could go wrong during connection closing, but let me move status check higher because status error seems to be more important than shutdown error.